addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-connection-service.js

```diff
@@ -360,6 +369,108 @@ class EnvironmentScConnectionService extends Service {
+        Condition: {
+          IpAddress: {
+            'aws:SourceIp': `${currentIpAddress.trim()}/32`,
+          },
+        },
+      };
+      const policyToUpdate = JSON.parse(decodeURIComponent(currentPolicyResponse.PolicyDocument));
+      let policyToUpdateStatement = policyToUpdate.Statement;
+      if (_.isArray(policyToUpdateStatement)) {
+        policyToUpdateStatement.push(newStatement);
+      } else {
+        policyToUpdateStatement = [policyToUpdateStatement, newStatement];
+        policyToUpdate.Statement = policyToUpdateStatement;
+      }
+      const putRolePolicyParams = {
+        RoleName: connection.role,
+        PolicyName: connection.policy,
+        PolicyDocument: JSON.stringify(policyToUpdate),
+      };
+      await iam.putRolePolicy(putRolePolicyParams).promise();
+
+      // wait for permission to propagate before using the role
+      await sleep(1000);
+      try {
+        const sageMakerResponseFn = async () => {
+          const stsEnvMgmt = await environmentScService.getClientSdkWithEnvMgmtRole(
+            requestContext,
+            { id: envId },
+            { clientName: 'STS', options: { apiVersion: '2017-07-24' } },
+          );
+          const {
+            Credentials: { AccessKeyId: accessKeyId, SecretAccessKey: secretAccessKey, SessionToken: sessionToken },
+          } = await stsEnvMgmt
+            .assumeRole({
+              RoleArn: connection.roleArn,
+              RoleSessionName: `create-presigned-url`,
+            })
+            .promise();
+          const sagemaker = await this.getSageMaker({ accessKeyId, secretAccessKey, sessionToken });
+          const params = {
+            NotebookInstanceName: connection.info,
+          };
+          const sageMakerResponse = await sagemaker.createPresignedNotebookInstanceUrl(params).promise();
+          return sageMakerResponse;
+        };
+        // Give sufficient number of retries to create presigned URL. This is needed because IAM role takes a while to
+        // update
+        const sageMakerResponse = await retry(sageMakerResponseFn, presign_retries);
+        return _.get(sageMakerResponse, 'AuthorizedUrl');
+      } catch (error) {
+        throw this.boom.internalError(`Could not generate presigned URL`, true).cause(error);
+      } finally {
+        // restore the original policy document. This ensures that caller IP address which was responsible for
+        // creating the presigned URL doesn't have access
+        const oldPolicyParams = { ...putRolePolicyParams };
+        oldPolicyParams.PolicyDocument = decodeURIComponent(currentPolicyResponse.PolicyDocument);
+        await iam.putRolePolicy(oldPolicyParams).promise();
+      }
+    });
+    return signedURL;
+  }
+
+  async getSageMaker({ accessKeyId, secretAccessKey, sessionToken }) {
+    const aws = await this.getAWS();
+    return new aws.sdk.SageMaker({ accessKeyId, secretAccessKey, sessionToken });
+  }
+
+  async getAWS() {
```

Comment:
We only use this method once, and it's only 2 lines. We might not need to create a method for this.

Reply:
This is done for making it easier to mock in unit tests.

Reply:
Added a comment